adapter/src/main/java/com/twilio/kudu/sql/rules/KuduSortRule.java

```diff
@@ -83,21 +84,24 @@ public boolean canApply(final RelTraitSet sortTraits, final KuduQuery query, fin
               && sortField.direction != RelFieldCollation.Direction.STRICTLY_ASCENDING)) {
         return false;
       }
-      // the sort columns must be a prefix of the primary key columns
-      if (sortField.getFieldIndex() >= openedTable.getSchema().getPrimaryKeyColumnCount()
-          || sortField.getFieldIndex() != pkColumnIndex) {
-        // This field is not in the primary key columns. If there is a condition lets
-        // see if it is there
-        if (filter.isPresent()) {
-          final RexNode originalCondition = filter.get().getCondition();
-          while (pkColumnIndex < sortField.getFieldIndex()) {
-            final KuduFilterVisitor visitor = new KuduFilterVisitor(pkColumnIndex);
-            final Boolean foundFieldInCondition = originalCondition.accept(visitor);
-            if (foundFieldInCondition.equals(Boolean.FALSE)) {
-              return false;
-            }
-            pkColumnIndex++;
+      if (sortField.getFieldIndex() >= openedTable.getSchema().getPrimaryKeyColumnCount()) {
+        return false;
+      }
+      if (sortField.getFieldIndex() != pkColumnIndex) {
+
+        if (predicates == null) {
```

Comment:
nit, remove check since `predicates` is initialized to null on line 68 and is not re-assigned until line 93?

Reply:
I believe calculating `mq.getAllPredicates` is expensive and we don't want to do it unless we need it. We could calculate it at the start (line 68), but if the Collations match the primary key columns or if the Collations do not contain a primary key, that calculation would never be used. 

So the idea is to start it as null on line 68. If the rule decides we need the filters, it checks if we have called `mq.getAllPredicates` here (line 92) and if not do the expansive work. If the expansive work comes up empty -- i.e. there are no filters -- return FALSE because we knew we needed filters but there isn't any below this `RelNode`. Otherwise, use the `predicates` to see if it matches the primary keys of Kudu.

This code I believe is correct and emulates a `lazy val` in Scala.

Reply:
With this [commit I have added comment for this](https://github.com/twilio/calcite-kudu/pull/2/commits/93000fbf4aebfb94cfdd2264fdaf3a8df96246ca#diff-bfe1fb131fa898e04d66af28c791f33c31b20586c6a2290308442915673c60a8R92)